gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/matching/single/cypher/operators/join/functions/MergeEmbeddings.java

```diff
@@ -36,56 +37,118 @@
  */
 public class MergeEmbeddings implements FlatJoinFunction<Embedding, Embedding, Embedding> {
   /**
-   * Join columns from the right side.
+   * Non-Join columns from the right side.
    */
-  private final List<Integer> joinColumnsRight;
+  private final int[] nonJoinColumnsRight;
+  /**
+   * Number of join columns of the right side.
+   */
+  private final int joinColumnsRightSize;
   /**
    * Vertex columns of the left embedding that need to have distinct id values.
    */
-  private final List<Integer> distinctVertexColumnsLeft;
+  private final int[] distinctVertexColumnsLeft;
   /**
    * Vertex columns of the right embedding that need to have distinct id values.
    */
-  private final List<Integer> distinctVertexColumnsRight;
+  private final int[] distinctVertexColumnsRight;
   /**
    * Edge columns of the left embedding that need to have distinct id values.
    */
-  private final List<Integer> distinctEdgeColumnsLeft;
+  private final int[] distinctEdgeColumnsLeft;
   /**
    * Edge columns of the right embedding that need to have distinct id values.
    */
-  private final List<Integer> distinctEdgeColumnsRight;
+  private final int[] distinctEdgeColumnsRight;
+  /**
+   * Flag, if vertex distinctiveness needs to be checked.
+   */
+  private final boolean checkDistinctVertices;
+  /**
+   * Flag, if vertex distinctiveness needs to be checked.
+   */
+  private final boolean checkDistinctEdges;
+  /**
+   * Reduce object instantiations.
+   */
+  private final Embedding reuseEmbedding;
 
   /**
    * Creates a new UDF instance.
    *
+   * @param rightColumns number of columns in the right embedding
    * @param joinColumnsRight join columns of the right side
    * @param distinctVertexColumnsLeft distinct vertex columns of the left embedding
    * @param distinctVertexColumnsRight distinct vertex columns of the right embedding
    * @param distinctEdgeColumnsLeft distinct edge columns of the left embedding
    * @param distinctEdgeColumnsRight distinct edge columns of the right embedding
    */
-  public MergeEmbeddings(List<Integer> joinColumnsRight,
+  public MergeEmbeddings(int rightColumns,
+    List<Integer> joinColumnsRight,
     List<Integer> distinctVertexColumnsLeft,
     List<Integer> distinctVertexColumnsRight,
     List<Integer> distinctEdgeColumnsLeft,
     List<Integer> distinctEdgeColumnsRight) {
-    this.joinColumnsRight           = joinColumnsRight;
-    this.distinctVertexColumnsLeft  = distinctVertexColumnsLeft;
-    this.distinctVertexColumnsRight = distinctVertexColumnsRight;
-    this.distinctEdgeColumnsLeft    = distinctEdgeColumnsLeft;
-    this.distinctEdgeColumnsRight   = distinctEdgeColumnsRight;
+
+    this.nonJoinColumnsRight = new int[rightColumns - joinColumnsRight.size()];
+    int idx = 0;
+    boolean isJoinColumn;
+    for (int i = 0; i < rightColumns; i++) {
```

Comment:
Instead of this a less complex solution would be
```java
    this.nonJoinColumnsRight = IntStream.range(0, rightColumns)
      .filter(pos -> !joinColumnsRight.contains(pos))
      .toArray();
```